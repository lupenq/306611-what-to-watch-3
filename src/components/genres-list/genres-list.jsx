import {connect} from "react-redux";
import {ActionCreator} from "../../reducer";
import MoviesList from "../movies-list/movies-list";
import ShowMore from '../show-more/show-more';

class GenresList extends React.PureComponent {
  constructor(props) {
    super(props);
  }

  getGenresList(filmsList) {
    return [`All genres`, ...new Set(filmsList.map((movie) => movie.genre))];
  }

  getMoviesByGenre(genre, movies) {
    return genre === `All genres` ? movies : movies.filter((movie) => movie.genre === genre);
  }

  render() {
    const {filmsList, genre, changeGenre, onMovieCardTitleClick, showingCardsNow} = this.props;

    return (
      <>
        <ul className="catalog__genres-list">
          {this.getGenresList(filmsList).map((activeGenre, index) => (
            <li
              className={`catalog__genres-item ${
                genre === activeGenre ? `catalog__genres-item--active` : ``
              }`}
              key={activeGenre + index}
            >
              <a
                className="catalog__genres-link"
                onClick={() => {
                  changeGenre(activeGenre);
                }}
              >
                {activeGenre}
              </a>
            </li>
          ))}
        </ul>

        <MoviesList
          filmsList={this.getMoviesByGenre(genre, filmsList)}
          onMovieCardTitleClick={onMovieCardTitleClick}
        />
        {
          this.getMoviesByGenre(genre, filmsList).length >= showingCardsNow && <ShowMore />
        }
      </>
    );
  }
}


GenresList.propTypes = {
  filmsList: PropTypes.array.isRequired,
  onMovieCardTitleClick: PropTypes.func.isRequired,
  genre: PropTypes.string.isRequired,
  changeGenre: PropTypes.func.isRequired,
  showingCardsNow: PropTypes.number.isRequired
};

const mapStateToProps = (state) => ({
  genre: state.genre,
  showingCardsNow: state.showingCardsNow,
  filmsList: state.films
});

const mapDispatchToProps = (dispatch) => ({
  changeGenre(genre) {
    dispatch(ActionCreator.changeGenre(genre));
  }
});

export {GenresList};
export default connect(mapStateToProps, mapDispatchToProps)(GenresList);
