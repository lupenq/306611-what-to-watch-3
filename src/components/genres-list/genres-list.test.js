import {GenresList} from "./genres-list";

const films = [
  {
    id: 1,
    name: `Интерстеллар1`,
    genre: `Фантастика`,
    year: `2014`,
    director: `Кристофер Нолан`,
    starring: `Мэттью МакКонахи, Энн Хэтэуэй, Джессика Честейн, Маккензи Фой, Майкл Кейн, Дэвид Гяси, Уэс Бентли, Мэтт Дэймон`,
    description: `Когда засуха, пыльные бури и вымирание растений приводят человечество к продовольственному кризису, коллектив исследователей и учёных отправляется сквозь червоточину (которая предположительно соединяет области пространства-времени через большое расстояние) в путешествие, чтобы превзойти прежние ограничения для космических путешествий человека и найти планету с подходящими для человечества условиями.`,
    bigPoster: `http://cinemaplex.ru/wp-content/uploads/2014/11/interstellar.jpg`,
    picture: `https://www.kinopoisk.ru/images/film_big/258687.jpg`,
    ratingScore: `2.5`,
    ratingLevel: `Very good`,
    ratingCount: `759`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  },
  {
    id: 2,
    name: `Зеленая книга`,
    genre: `комедия`,
    year: `2018`,
    director: `Питер Фаррелли`,
    starring: `Вигго Мортенсен, Махершала Али, Линда Карделлини, Себастьян Манискалко, Димитар Маринов, Майк Хаттон`,
    description: `1960-е годы. После закрытия нью-йоркского ночного клуба на ремонт вышибала Тони по прозвищу Болтун ищет подработку на пару месяцев. Как раз в это время Дон Ширли, утонченный светский лев, богатый и талантливый чернокожий музыкант, исполняющий классическую музыку, собирается в турне по южным штатам, где ещё сильны расистские убеждения и царит сегрегация. Он нанимает Тони в качестве водителя, телохранителя и человека, способного решать текущие проблемы. У этих двоих так мало общего, и эта поездка навсегда изменит жизнь обоих.`,
    bigPoster: `https://cinemaholics.ru/content/images/2018/08/green-book_universal-1920x1181.jpg`,
    picture: `https://www.kinopoisk.ru/images/film_big/1108577.jpg`,
    ratingScore: `10`,
    ratingLevel: `Very good`,
    ratingCount: `2566`,
    preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  }];

it(`Should render GenresList`, () => {
  const tree = renderer
    .create(
        <GenresList
          filmsList={films}
          genre={`All genres`}
          changeGenre={() => {}}
          onMovieCardTitleClick={() => {}}
        />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
