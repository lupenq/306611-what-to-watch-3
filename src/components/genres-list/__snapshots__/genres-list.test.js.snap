// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render GenresList 1`] = `
Array [
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item catalog__genres-item--active"
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        All genres
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Фантастика
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        комедия
      </a>
    </li>
  </ul>,
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <img
          alt="Интерстеллар1"
          height="175"
          src="https://www.kinopoisk.ru/images/film_big/258687.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="movie-page.html"
          onClick={[Function]}
        >
          Интерстеллар1
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <img
          alt="Зеленая книга"
          height="175"
          src="https://www.kinopoisk.ru/images/film_big/1108577.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="movie-page.html"
          onClick={[Function]}
        >
          Зеленая книга
        </a>
      </h3>
    </article>
  </div>,
]
`;
